bat/workload.go

```diff
@@ -0,0 +1,196 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Package bat contains a number of helper functions that can be used to perform
+// various operations on a ciao cluster such as creating an instance or retrieving
+// a list of all the defined workloads, etc.  All of these helper functions are
+// implemented by calling ciao-cli rather than by using ciao's REST APIs.  This
+// package is mainly intended for use by BAT tests.  Manipulating the cluster
+// via ciao-cli, rather than through the REST APIs, allows us to test a little
+// bit more of ciao.
+package bat
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+// Source is provided to the disk structure to indicate whether the
+// disk should be cloned from a volume or an image.
+type Source struct {
+	Type string `yaml:"service"`
+	ID   string `yaml:"id"`
+}
+
+// Disk describes the storage for the workload definition.
+type Disk struct {
+	ID        *string `yaml:"volume_id"`
+	Size      int     `yaml:"size"`
+	Bootable  bool    `yaml:"bootable"`
+	Source    *Source `yaml:"source"`
+	Ephemeral bool    `yaml:"ephemeral"`
+}
+
+// DefaultResources indicate how many cpus and mem to allocate.
+type DefaultResources struct {
+	VCPUs int `yaml:"vcpus"`
+	MemMB int `yaml:"mem_mb"`
+}
+
+// WorkloadOptions is used to generate a workload definition in yaml.
+type WorkloadOptions struct {
+	Description     string           `yaml:"description"`
+	VMType          string           `yaml:"vm_type"`
+	FWType          string           `yaml:"fw_type"`
+	ImageName       string           `yaml:"image_id"`
+	Defaults        DefaultResources `yaml:"defaults"`
+	CloudConfigFile string           `yaml:"cloud_init"`
+	Disks           []Disk           `yaml:"disks"`
+}
+
+// this function should output the yaml file for the workload definition
+func createWorkloadDefinition(filename string, opt WorkloadOptions) error {
+	// take the cloud config file and output it to a file.
+	f, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	y, err := yaml.Marshal(opt)
+	if err != nil {
+		return err
+	}
+
+	_, err = f.Write(y)
+	return err
+}
+
+// this function should output a cloud init file for the workload definition
+func createCloudConfig(filename string, config string) error {
+	// take the cloud config file and output it to a file.
+	f, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	_, err = f.WriteString(config)
+	return err
+}
+
+func createWorkload(ctx context.Context, tenant string, opt WorkloadOptions, config string, public bool) (string, error) {
+	// write out the cloud config.
+	err := createCloudConfig("config.yaml", config)
```

Comment:
It might be better to write these files to somewhere under /tmp to avoid cluttering the working directory (in the case of a crash for example).  There are functions in ioutil for creating temporary files and directories.